core/src/main/java/edu/ucr/cs/riple/core/metadata/submodules/DownStreamDependencyAnalyzer.java

```diff
@@ -0,0 +1,170 @@
+import edu.ucr.cs.riple.core.explorers.OptimizedExplorer;
+import edu.ucr.cs.riple.core.injectors.VirtualInjector;
+import edu.ucr.cs.riple.core.metadata.field.FieldDeclarationAnalysis;
+import edu.ucr.cs.riple.core.metadata.index.Bank;
+import edu.ucr.cs.riple.core.metadata.index.Error;
+import edu.ucr.cs.riple.core.metadata.index.Fix;
+import edu.ucr.cs.riple.core.metadata.method.MethodInheritanceTree;
+import edu.ucr.cs.riple.core.metadata.method.MethodNode;
+import edu.ucr.cs.riple.core.metadata.trackers.MethodRegionTracker;
+import edu.ucr.cs.riple.core.util.Utility;
+import edu.ucr.cs.riple.injector.changes.AddAnnotation;
+import edu.ucr.cs.riple.injector.location.OnMethod;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Analyzer for downstream dependencies. It collects public APIs used in the dependencies and
+ * collects a summarization of an overall effect across all downstream dependencies. These data can
+ * be fed to the Annotator main process in the decision process.
+ */
+public class DownStreamDependencyAnalyzer {
+
+  /** Set of downstream dependencies. */
+  private final ImmutableSet<Module> modules;
+  /** Public APIs in the target modules that have a non-primitive return value. */
+  private final Stream<MethodStatus> methods;
+  /** Annotator Config. */
+  private final Config config;
+  /** Method inheritance instance. */
+  private final MethodInheritanceTree tree;
+  /**
+   * VirtualInjector instance, annotations are not going to be actually injected physically in the
+   * source codes and should only be loaded in a library model.
+   */
+  private final VirtualInjector injector;
+
+  /** Container class for storing overall effect of each method. */
+  static class MethodStatus {
+    /** Node in {@link MethodInheritanceTree }corresponding to a public method. */
+    final MethodNode node;
+    /**
+     * Effect of injecting a {@code Nullable} annotation on pointing method of node on downstream
+     * dependencies.
+     */
+    int effect;
+
+    public MethodStatus(MethodNode node) {
+      this.node = node;
+      this.effect = 0;
+    }
+  }
+
+  public DownStreamDependencyAnalyzer(Config config, MethodInheritanceTree tree) {
+    this.config = config;
+    this.modules = config.getSubModules();
+    this.tree = tree;
+    this.injector = new VirtualInjector(config);
+    this.methods = tree.getPublicMethodsWithNonPrimitivesReturn().stream().map(MethodStatus::new);
+  }
+
+  /**
+   * Exploration phase begins. When this method finishes, all methods effect on downstream
+   * dependencies are calculated.
+   */
+  public void explore() {
+    modules.forEach(this::analyzeDownstreamDependency);
+  }
+
+  /**
+   * Analyzes effects of public methods on a module.
+   *
+   * @param module A downstream dependency.
+   */
+  private void analyzeDownstreamDependency(Module module) {
+    Utility.setScannerCheckerActivation(config, true);
+    Utility.buildProject(config, module);
```

Comment:
One thing to note is that we are working under a model that has us building all downstream dependencies one at a time. This is fine for now, but note that it's way more efficient to build all direct dependencies of a target in parallel, and it shouldn't cause any issues, because - since they are _direct_ downstream dependencies - none of those targets will depend on each other.

I am wondering a bit how that will work with this runner. Presumably it's as easy as passing a single build command with N targets and this will treat it as one module but otherwise work as expected? Or do we rely on the fact that we control building _specific_ downstream dependencies one at a time?

Reply:
It depends on the error prone flag values we are passing to downstream modules. One of the `Scanner` flag value is the path to the config file where it sets the output directory for all `Scanner` outputs. If we set a different value for each downstream dependency we can run this in parallel provided we inform `Annotator` about different output dir paths.

Reply:
For potential observers / other reviewers: This turned out to be a longer design discussion, concluded offline. We plan to support building multiple targets in parallel with a single build system command, but this requires some changes on a follow PR, since this one grows quite large 😄